protos/sigstore_common.proto

```diff
@@ -0,0 +1,105 @@
+// UNSPECIFIED SHOULD not be used, primary reason for inclusion is to force
+// any proto JSON serialization to emit the used hash algorithm, as default
+// option is to *omit* the default value of an emum (which is the first
+// value, represented by '0'.
+enum HashAlgorithm {
+        UNSPECIFIED = 0;
+        SHA2_256 = 1;
+        SHA2_512 = 2;
+}
+
+// HashOutput captures a digest of a 'message' (generic octet sequence)
+// and the corresponding hash algorithm used.
+message HashOutput {
+        HashAlgorithm algorithm = 1;
+        // This is the raw octets of the message digest as computed by
+        // the hash algorithm.
+        bytes digest = 2;
+}
+
+// MessageSignature stores the computed signature over a message.
+message MessageSignature {
+        // Message digest can be used to identify the artifact.
+        HashOutput message_digest = 1;
+        // The raw bytes as returned from the signature algorithm.
+        // The signature algorithm (and so the format of the signature bytes)
+        // are determined by the contents of the 'verification_material',
+        // either a key-pair or a certificate. If using a certificate, the
+        // certificate contains the required information on the signature
+        // algorithm.
+        // When using a key pair, the algorithm MUST be part of the public
+        // key, which MUST be communicated out-of-band.
+        bytes signature = 2;
+}
+
+// This message holds a RFC 3161 timestamp.
+message RFC3161SignedTimestamp {
+        // Signed timestamp is the DER encoded TimeStampToken
+        // See https://www.rfc-editor.org/rfc/rfc3161.html#section-2.4.2
+        // For verification, the TimeStampToken MUST contain TSA's
+        // certificate chain.
+        bytes signed_timestamp = 1;
+}
+
+
+// PublicKeyIdentifier can be used to identify an (out of band) delivered
+// key, to verify a signature.
+message PublicKeyIdentifier {
+        // Optional unauthenticated hint on which key to use.
+        // The format of the hint must be agreed upon out of band by the
+        // signer and the verifiers, and so is not subject to this
+        // specification.
+        string hint = 1;
+}
+
+message X509Certificate {
+        // DER encoded X509 certificate.
+        bytes der_bytes = 1;
+}
+
+// A chain of X509 certificates.
+message X509CertificateChain {
+        // The chain of certificates, with indices 0 to n.
+        // The first certificate in the array  must be the leaf
+        // certificate used for signing. Any intermediate certificates
+        // must be stored as offset 1 to n-1, and the root certificate at
+        // position n.
+        repeated X509Certificate certificates = 1;
+}
+
+// VerificationMaterial captures details on the materials used to verify
+// signatures.
+message VerificationMaterial {
+        oneof content {
+                PublicKeyIdentifier public_key = 1;
+                X509CertificateChain x509_certificate_chain = 2;
+        }
+}
```

Comment:
Ah, something that I think may be missing for Fulcio verification is the possibility of a detached SCT header. The public good Fulcio instance will always provide an embedded SCT (promise of inclusion from the CT log). However, private deployments may either specify a detached SCT or exclude CT entirely (which is not ideal, but depends on the set up). I think we need a lightweight sigstore_fulcio.proto (or at least a message in common.proto) that provides a certificate chain and an optional detached SCT - like https://github.com/sigstore/fulcio/blob/main/fulcio.proto#L153  

To abstract away Fulcio entirely, we could just wrap `X509CertificateChain` in another message, such as:

```
message SignedCertificateTimestamp {
  bytes sct = 1;
}

message CertificateVerifier {
  X509CertificateChain x509_certificate_chain = 1;
  SignedCertificateTimestamp sct = 2;
}

message VerificationMaterial {
        oneof content {
                PublicKeyIdentifier public_key = 1;
                CertificateVerifier certificate_verifier = 2;
        }
}
```

or as its own proto, but I like the idea that Fulcio isn't a requirement, it's just that you need a certificate chain (from fulcio or any CA) and an SCT (from fulcio or your own CT log)

See https://github.com/sigstore/cosign/pull/1731 or https://github.com/sigstore/cosign/issues/2382 for some comments on it.

cc @znewman01

Reply:
Alternatively, we take the stance that you are never allowed to provide a detached SCT, and it must always be embedded, but this would mean updating Fulcio and dropping certain signing backends

also cc @bobcallaway

Reply:
I don't think we should depend on Fulcio at all, so +1 on that. On supporting detached SCT or not, I can't tell, but supporting them seems good. From a verifier view, what happens if the certificate has an SCT embedded, and one detached? I think this is too big of a question to answer _now_. I would suggest go with embedded SCTs now, then after we have merged this, open up an issue on this topic. I already have a list of issues that I will open when we have the first version merged.

Reply:
SG, happy to address after this is merged.

That is a great question and one that came up when discussing detached SCTs previously too. This is why we structured Fulcio's response to return either a cert or a cert+SCT - Maybe we have the same structure here.